src/Unleash/Utilities/ToggleBootstrapUrlProvider.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Unleash.Internal;
+
+namespace Unleash.Utilities
+{
+    public class ToggleBootstrapUrlProvider : IToggleBootstrapProvider
+    {
+        private readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+        private readonly HttpClient client;
+        private readonly string path;
+        private readonly bool throwOnFail;
+
+        public ToggleBootstrapUrlProvider(string path, bool throwOnFail = false)
+        {
+            this.path = path;
+            client = new HttpClient();
+            this.throwOnFail = throwOnFail;
+        }
+
+        public ToggleBootstrapUrlProvider(string path, HttpClient client, bool throwOnFail = false)
+        {
+            this.path = path;
+            this.client = client;
+            this.throwOnFail = throwOnFail;
+        }
+
+        public string Read()
+        {
+            return Task.Run(() => FetchFile()).Result;
```

Comment:
Seems like a terrible shame to go through the async/await dance only to synchronize and block. Is this so we can use this in a constructor?

Reply:
Yeah, this is done in the UnleashServices constructor, and while we could probably move this all the way up to the constructor and keep it async down on from there i didn't want to casually introduce that big a change. The other thing i took into consideration is that this most likely occurs only once during startup, and freeing up threads while waiting on IO might not be a priority at that point

Reply:
Yeah I hear you, that sounds like a rather big change. I'm not very familiar with this SDK and it seems to want to do things in different ways from the others. I think we should just accept this and potentially look at a refactor after (if we even want to, I think your arguments for leaving it alone are good)

Reply:
Task.Run(() => FooBar()) Should be used in a CPU bound operation not in a I/O Bound operation, also, is better to use GetAwaiter().GetResult() than .Result directly because in case of failure, the first one will propagate the exceptions and the second one will wrap in a AggregateException.

Reply:
@amj1985 Thank you, yes indeed. The reason i opted for Task.Run to avoid deadlocks. I'll admit to not being the ultimate brains on this topic, but my understanding is that GetAwaiter().GetResult() can cause deadlocks in things like ASP.NET (not Core) and other single threaded use cases that can't neccessarily be ruled out

Reply:
Both cases .Result .GetAwaiter().GetResult() cause the potential for deadlock and thread pool starvation issues. It should be avoided and be implemented using async all the way, but it involves a big refactor, my recommendation is to use GetAwaiter().GetResult(), if the process throws
a FileNotFoundException will get the actual exception instead of an AggregateException

Reply:
@amj1985 ah i see what you mean now, apologies, i missread. Task.Run(...).GetAwaiter().GetResult() vs .Result. Yeah reasonable request, i'll fix that!

Reply:
@amj1985 I think we should lean towards discussing a refactor soon. The fact that this is currently hard to implement without dissecting technical internals of dotnet runtimes suggests that its time! (I'm keen and willing to help here)